File: data/examples/speculative-generality.cpp
// 过度设计的抽象层次
template<typename T>
class AbstractDataProcessor {
public:
    virtual void process(T data) = 0;
    virtual void preProcess(T data) {}
    virtual void postProcess(T data) {}
    virtual bool shouldProcess(T data) { return true; }
    virtual void onError(const std::exception& e) {}
    virtual void onSuccess(T data) {}
    virtual ~AbstractDataProcessor() = default;
};

template<typename T>
class ConfigurableDataProcessor : public AbstractDataProcessor<T> {
private:
    std::map<std::string, std::function<void(T)>> processors;
    std::vector<std::function<bool(T)>> filters;
    std::vector<std::function<void(T)>> preprocessors;
    std::vector<std::function<void(T)>> postprocessors;
    
public:
    void addProcessor(const std::string& name, std::function<void(T)> proc) {
        processors[name] = proc;
    }
    
    void addFilter(std::function<bool(T)> filter) {
        filters.push_back(filter);
    }
    
    void addPreprocessor(std::function<void(T)> preprocessor) {
        preprocessors.push_back(preprocessor);
    }
    
    void process(T data) override {
        // 复杂的处理逻辑，但实际上只有一种使用方式
        for (auto& preprocessor : preprocessors) {
            preprocessor(data);
        }
        
        for (auto& filter : filters) {
            if (!filter(data)) return;
        }
        
        for (auto& [name, processor] : processors) {
            processor(data);
        }
    }
};

// 实际上只需要这样简单的实现
class SimpleStringProcessor {
public:
    void process(const std::string& data) {
        std::cout << "Processing: " << data << std::endl;
    }
};

File: data/examples/speculative-generality.java
import java.util.*;
import java.util.function.*;

// Overly designed abstraction layers
abstract class AbstractDataProcessor<T> {
    public abstract void process(T data);
    public void preProcess(T data) {}
    public void postProcess(T data) {}
    public boolean shouldProcess(T data) { return true; }
    public void onError(Exception e) {}
    public void onSuccess(T data) {}
}

class ConfigurableDataProcessor<T> extends AbstractDataProcessor<T> {
    private Map<String, Consumer<T>> processors;
    private List<Predicate<T>> filters;
    private List<Consumer<T>> preprocessors;
    private List<Consumer<T>> postprocessors;
    
    public ConfigurableDataProcessor() {
        this.processors = new HashMap<>();
        this.filters = new ArrayList<>();
        this.preprocessors = new ArrayList<>();
        this.postprocessors = new ArrayList<>();
    }
    
    public void addProcessor(String name, Consumer<T> processor) {
        processors.put(name, processor);
    }
    
    public void addFilter(Predicate<T> filter) {
        filters.add(filter);
    }
    
    public void addPreprocessor(Consumer<T> preprocessor) {
        preprocessors.add(preprocessor);
    }
    
    @Override
    public void process(T data) {
        // Complex processing logic, but actually only used in one way
        for (Consumer<T> preprocessor : preprocessors) {
            preprocessor.accept(data);
        }
        
        for (Predicate<T> filter : filters) {
            if (!filter.test(data)) return;
        }
        
        for (Consumer<T> processor : processors.values()) {
            processor.accept(data);
        }
    }
}

// What we actually need - a simple implementation
class SimpleStringProcessor {
    public void process(String data) {
        System.out.println("Processing: " + data);
    }
}
